From 498a30a204f98831547dbe07d1d59f615dbef4c1 Mon Sep 17 00:00:00 2001
From: Damien George <damien@micropython.org>
Date: Sat, 17 Oct 2026 15:21:48 -0700
Subject: [PATCH] mimxrt/modmachine: Fix machine.reset_cause()
 (micropython#19619).

micropython#19619's three commits in one, rebased onto openmv's fork:
- mimxrt/modmachine: Expose machine.HARD_RESET and DEEPSLEEP_RESET consts.
- mimxrt/modmachine: Get machine.SOFT_RESET and HARD_RESET working.
- mimxrt/modmachine: Fix machine.DEEPSLEEP_RESET reset cause.

Fork compat: also declares machine_rtc_alarm_off() in modmachine.h, which
upstream did in an earlier commit (09f5bb4) the fork does not have.
---
 ports/mimxrt/main.c       |  2 ++
 ports/mimxrt/modmachine.c | 59 ++++++++++++++++++++++++++++-----------
 ports/mimxrt/modmachine.h |  4 +++
 3 files changed, 48 insertions(+), 17 deletions(-)

diff --git a/ports/mimxrt/main.c b/ports/mimxrt/main.c
index 05a3ed82d..020a2209b 100644
--- a/ports/mimxrt/main.c
+++ b/ports/mimxrt/main.c
@@ -72,6 +72,7 @@ int main(void) {
     board_init();
     ticks_init();
     pendsv_init();
+    machine_init();
 
     #if MICROPY_HW_ENABLE_PSRAM
     size_t psram_size = configure_external_ram();
@@ -206,6 +207,7 @@ int main(void) {
         #if MICROPY_PY_MACHINE_QECNT
         machine_encoder_deinit_all();
         #endif
+        machine_set_soft_reset();
         gc_sweep_all();
         mp_deinit();
     }
diff --git a/ports/mimxrt/modmachine.c b/ports/mimxrt/modmachine.c
index 452d045f6..490c428fb 100644
--- a/ports/mimxrt/modmachine.c
+++ b/ports/mimxrt/modmachine.c
@@ -68,7 +68,9 @@
     \
     /* Reset reasons */ \
     { MP_ROM_QSTR(MP_QSTR_PWRON_RESET),         MP_ROM_INT(MP_PWRON_RESET) }, \
+    { MP_ROM_QSTR(MP_QSTR_HARD_RESET),          MP_ROM_INT(MP_HARD_RESET) }, \
     { MP_ROM_QSTR(MP_QSTR_WDT_RESET),           MP_ROM_INT(MP_WDT_RESET) }, \
+    { MP_ROM_QSTR(MP_QSTR_DEEPSLEEP_RESET),     MP_ROM_INT(MP_DEEPSLEEP_RESET) }, \
     { MP_ROM_QSTR(MP_QSTR_SOFT_RESET),          MP_ROM_INT(MP_SOFT_RESET) }, \
 
 typedef enum {
@@ -84,6 +86,46 @@ typedef enum {
 #define DBL_TAP_MAGIC 0xf01669ef // Randomly selected, adjusted to have first and last bit set
 #define DBL_TAP_MAGIC_QUICK_BOOT 0xf02669ef
 
+static uint8_t reset_cause;
+
+void machine_init(void) {
+    #ifdef MIMXRT117x_SERIES
+    const uint32_t user_reset_flag = kSRC_M7CoreIppUserResetFlag;
+    #else
+    const uint32_t user_reset_flag = kSRC_IppUserResetFlag;
+    #endif
+
+    // machine.deepsleep() uses SNVS_LPCR_TOP_MASK to turn off power, which sets
+    // SNVS_LPSR_EO_MASK.  Use that bit to tell if we are waking from deepsleep.
+    bool woke_from_deepsleep = !!(SNVS->LPSR & SNVS_LPSR_EO_MASK);
+    SNVS->LPSR = SNVS_LPSR_EO_MASK;
+
+    // Check if the device was reset due to low-power timer alarm.
+    if (SNVS->LPSR & SNVS_LPSR_LPTA_MASK) {
+        machine_rtc_alarm_off(true);
+        woke_from_deepsleep = true;
+    }
+
+    // Determine the current reset cause.
+    if (woke_from_deepsleep || (SRC->SRSR & user_reset_flag)) {
+        reset_cause = MP_DEEPSLEEP_RESET;
+    } else {
+        uint16_t wdog =
+            WDOG_GetStatusFlags(WDOG1) & (kWDOG_PowerOnResetFlag | kWDOG_TimeoutResetFlag | kWDOG_SoftwareResetFlag);
+        if (wdog == kWDOG_PowerOnResetFlag) {
+            reset_cause = MP_PWRON_RESET;
+        } else if (wdog == kWDOG_TimeoutResetFlag) {
+            reset_cause = MP_WDT_RESET;
+        } else {
+            reset_cause = MP_HARD_RESET;
+        }
+    }
+}
+
+void machine_set_soft_reset(void) {
+    reset_cause = MP_SOFT_RESET;
+}
+
 static mp_obj_t mp_machine_unique_id(void) {
     unsigned char id[8];
     mp_hal_get_unique_id(id);
@@ -98,23 +140,6 @@ MP_NORETURN static void mp_machine_reset(void) {
 }
 
 static mp_int_t mp_machine_reset_cause(void) {
-    #ifdef MIMXRT117x_SERIES
-    uint32_t user_reset_flag = kSRC_M7CoreIppUserResetFlag;
-    #else
-    uint32_t user_reset_flag = kSRC_IppUserResetFlag;
-    #endif
-    if (SRC->SRSR & user_reset_flag) {
-        return MP_DEEPSLEEP_RESET;
-    }
-    uint16_t reset_cause =
-        WDOG_GetStatusFlags(WDOG1) & (kWDOG_PowerOnResetFlag | kWDOG_TimeoutResetFlag | kWDOG_SoftwareResetFlag);
-    if (reset_cause == kWDOG_PowerOnResetFlag) {
-        reset_cause = MP_PWRON_RESET;
-    } else if (reset_cause == kWDOG_TimeoutResetFlag) {
-        reset_cause = MP_WDT_RESET;
-    } else {
-        reset_cause = MP_SOFT_RESET;
-    }
     return reset_cause;
 }
 
diff --git a/ports/mimxrt/modmachine.h b/ports/mimxrt/modmachine.h
index 8a26d3ee6..c1486b51e 100644
--- a/ports/mimxrt/modmachine.h
+++ b/ports/mimxrt/modmachine.h
@@ -32,6 +32,9 @@
 extern const mp_obj_type_t machine_can_type;
 extern const mp_obj_type_t machine_sdcard_type;
 
+void machine_init(void);
+void machine_set_soft_reset(void);
+
 void machine_adc_init(void);
 void machine_can_irq_deinit(void);
 void machine_pin_irq_deinit(void);
@@ -43,6 +46,7 @@ void machine_sdcard_init0(void);
 void mimxrt_sdram_init(void);
 void machine_i2s_init0();
 void machine_i2s_deinit_all(void);
+void machine_rtc_alarm_off(bool clear);
 void machine_rtc_start(void);
 uint64_t machine_rtc_get_ticks(void);
 uint64_t machine_rtc_get_seconds(void);
